app/helpers/projects_helper.rb

```diff
@@ -1,2 +1,134 @@
+  def validate_metadata_csv_for_project(project_samples, metadata, new_samples)
+    errors = []
+    warnings = []
+
+    if new_samples
+      # Require host_genome_name as a column.
+      unless metadata["headers"].include?("host_genome_name")
+        errors.push("host_genome_name column is required.")
+        return { "errors" => errors, "warnings" => warnings }
+      end
+    else
+      # Require sample_name as a column.
+      unless metadata["headers"].include?("sample_name")
+        errors.push("sample_name column is required.")
+        return { "errors" => errors, "warnings" => warnings }
+      end
+    end
+
+    # Verify that the column names are supported.
+    metadata["headers"].each_with_index do |header, index|
+      if header != "sample_name" && header != "host_genome_name" && !Metadatum::KEY_TO_TYPE.key?(header.to_sym)
+        errors.push("#{header} (column #{index + 1}) is not a supported metadata type.")
+      end
+    end
+
+    sample_name_index = metadata["headers"].find_index("sample_name")
+    host_genome_name_index = metadata["headers"].find_index("host_genome_name")
+
+    metadata["rows"].each_with_index do |row, index|
+      # Check number of values in the row.
+      if row.length > metadata["headers"].length
+        errors.push("Row #{index} has too many values. (#{row.length} instead of #{metadata['headers'].length})")
+      end
+      if row.length < metadata["headers"].length
+        errors.push("Row #{index} has too few values. (#{row.length} instead of #{metadata['headers'].length})")
+      end
+
+      if new_samples
+        # Check for valid host_genome_name and create temporary sample
+        if row[host_genome_name_index].nil? || row[host_genome_name_index] == ""
+          errors.push("Row #{index} is missing host_genome_name.")
+          next
+        end
+        host_genome = HostGenome.where(name: row[host_genome_name_index]).first
+        if host_genome.nil?
+          errors.push("#{row[host_genome_name_index]} (row #{index}) is an invalid host_genome_name.")
+          next
+        end
+
+        # Create temporary sample to run metadatum validation
+        sample = Sample.new
+        sample.host_genome_id = host_genome.id
+      else
+        # Check for valid sample name and fetch sample
+        if row[sample_name_index].nil? || row[sample_name_index] == ""
+          errors.push("Row #{index} is missing sample_name.")
+          next
+        end
+        sample = project_samples.where(name: row[sample_name_index]).first
+        if sample.nil?
+          errors.push("#{row[sample_name_index]} (row #{index}) does not match any sample names in this project.")
+          next
+        end
+      end
+
+      # Validate the metadatum values with the sample.
+      row.each_with_index do |value, row_index|
+        if row_index >= metadata["headers"].length
+          next
+        end
```

Comment:
FYI there's a nice way you can write `next if row_index >= metadata["headers"].length` to make the code a bit more concise